greenmail-core/src/test/java/com/icegreen/greenmail/test/specificmessages/EncodingTest.java

```diff
@@ -189,6 +189,52 @@ public void messagesAdded(MessageCountEvent e) {
         }
     }
 
+    @Test(timeout = 10000)
+    public void testTextPlainWithUTF8SubjectAndGreenMailApi() throws MessagingException, IOException {
+        greenMail.setUser("to@localhost", "pwd");
+        final IMAPStore store = greenMail.getImap().createStore();
+        store.connect("to@localhost", "pwd");
+        try {
+            String subject = "кирилица ünicöde_\uD83C\uDF36";
+            Folder inboxFolder = store.getFolder("INBOX");
+            inboxFolder.open(Folder.READ_ONLY);
+            Message[] messages = new Message[] { null };
+            MessageCountListener listener = new MessageCountListener() {
+                @Override
+                public void messagesRemoved(MessageCountEvent e) {
+                }
+
+                @Override
+                public void messagesAdded(MessageCountEvent e) {
+                    messages[0] = e.getMessages()[0];
+                }
+            };
+            inboxFolder.addMessageCountListener(listener);
+            new Thread(() -> {
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e1) {
+                    // Ignore
+                }
+                try {
+                    MimeMessage message = GreenMailUtil.createTextEmail(
+                        "to@localhost", "from@localhost", "will be replaced", "body", greenMail.getSmtp().getServerSetup());
+                    message.setHeader("Subject", subject);
```

Comment:
All headers must be encoded, see https://www.rfc-editor.org/rfc/rfc2047 for details.

Could be done e.g. via
```
message.setHeader("Subject", MimeUtility.encodeText(subject));
```
or indirectly by using
```
message.setSubject(...)
```

Longer subjects (encoded >76 chars) need to be folded, too. Example from https://github.com/jakartaee/mail-api/blob/master/api/src/main/java/jakarta/mail/internet/MimeMessage.java#L874 :
```
		setHeader("Subject", MimeUtility.fold(9,
		    MimeUtility.encodeText(subject, charset, null)));
```

Reply:
@marcelmay thanks for the response. I obviously haven't read the RFC, however isn't the `mail.mime.allowutf8` property exactly for that? Doesn't it allow setting UTF-8 symbols in headers without encoding first? That requires server support as far as I understand but otherwise it's supposed to work. Or did I misunderstand something? If I'm correct, then this test should work (after `mail.mime.allowutf8` is somehow set - which I haven't done - please see my initial comment).

Reply:
After diving into JavaMail sources this will only work if the SMTP server also announces SMTPUTF8, which GreenMail currently does not support:

https://github.com/javaee/javamail/blob/master/mail/src/main/java/com/sun/mail/smtp/SMTPTransport.java#L768 :
```
	    if (allowutf8 && !supportsExtension("SMTPUTF8"))
		logger.log(Level.INFO, "mail.mime.allowutf8 set " +
			    "but server doesn't advertise SMTPUTF8 support");
```

It is really mail server dependent. If I force the property, it shows up in the logs:
```
Feb 06, 2023 9:55:41 PM com.sun.mail.smtp.SMTPTransport protocolConnect
INFO: mail.mime.allowutf8 set but server doesn't advertise SMTPUTF8 support
```

I created #535 for SMTPUTF8. Though I believe not many servers support it ... ?

Reply:
No idea how many servers support it. And yes, I guess GreenMail also doesn't support that. But in any case, this is rather low-priority I would think. #481 and #483 are much more important.